File: src/main/java/io/github/chaosdave34/kitpvp/enchantments/FreezeEnchantment.java
package io.github.chaosdave34.kitpvp.enchantments;

import io.github.chaosdave34.ghutils.enchantment.CustomEnchantment;
import net.minecraft.world.entity.EquipmentSlot;
import net.minecraft.world.item.enchantment.EnchantmentCategory;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

public class FreezeEnchantment extends CustomEnchantment {
    public FreezeEnchantment() {
        super("freeze", 3, "Freeze", EnchantmentCategory.WEAPON, EquipmentSlot.MAINHAND);
    }

    @EventHandler
    public void onAttack(EntityDamageByEntityEvent e) {
        if (e.getCause() == EntityDamageEvent.DamageCause.ENTITY_ATTACK) {
            if (e.getDamager() instanceof Player damager) {
                ItemStack itemStack = damager.getInventory().getItemInMainHand();
                if (itemStack.containsEnchantment(CustomEnchantments.FREEZE)) {
                    int lvl = itemStack.getEnchantmentLevel(CustomEnchantments.FREEZE);
                    if (e.getEntity() instanceof LivingEntity entity) {
                        entity.addPotionEffect(new PotionEffect(PotionEffectType.SLOW, lvl * 20, 1, false, false));
                    }
                }
            }
        }
    }
}

File: src/main/kotlin/io/github/chaosdave34/kitpvp/enchantments/impl/FreezeEnchantment.kt
package io.github.chaosdave34.kitpvp.enchantments.impl

import io.github.chaosdave34.kitpvp.enchantments.CustomEnchantment
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.entity.EntityDamageEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType

class FreezeEnchantment : CustomEnchantment("freeze", "Freeze") {
    @EventHandler
    fun onAttack(event: EntityDamageByEntityEvent) {
        if (event.cause == EntityDamageEvent.DamageCause.ENTITY_ATTACK) {
            val player = event.damager
            if (player is Player) {
                val itemStack: ItemStack = player.inventory.itemInMainHand
                if (itemStack.containsThisEnchantment()) {
                    val lvl = itemStack.getEnchantmentLevel()
                    if (event.entity is LivingEntity) {
                        player.addPotionEffect(PotionEffect(PotionEffectType.SLOWNESS, lvl * 20, 1, false, false))
                    }
                }
            }
        }
    }
}